pkg/meta/distributed_lock.go

```diff
@@ -0,0 +1,310 @@
+				return nil, nil
+			}
+
+			// Lock is exist, update it.
+			lc.StartTS = txn.StartTS()
+			data, err := lc.encode()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			err = txn.Set(d.lockName, data)
+			if err != nil {
+				return errors.Trace(err), nil
+			}
+			return nil, nil
+		})
+}
+
+var lockHeldErr = errors.Errorf("lock is held by another instance")
+
+// Lock locks the distributed lock.
+func (d *DistributedLock) Lock() error {
+	err := d.getLockInTxnWithRetry(d.outerCtx,
+		func(lc *lockContent, txn DataTxn) (storeErr, otherErr error) {
+			startTS := txn.StartTS()
+			// Lock is not exist or expired, create or overwrite it.
+			if lc == nil || d.lockExpired(lc, startTS) {
+				lc := newLockContent(startTS, d.instanceID)
+				data, err := lc.encode()
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				err = txn.Set(d.lockName, data)
+				if err != nil {
+					return errors.Trace(err), nil
+				}
+				return nil, nil
+			}
+			// Lock is valid, wait for a while.
+			return lockHeldErr, nil
+		})
+	if err != nil {
+		return err
+	}
+	if d.lease > 0 {
+		var renewCtx context.Context
+		renewCtx, d.cancelRenewLoop = context.WithCancel(d.outerCtx)
+		go d.renewLockLoop(renewCtx)
+	}
+	return nil
+}
+
+func (d *DistributedLock) lockExpired(lc *lockContent, currentStartTS uint64) bool {
+	if d.lease == 0 {
+		return false
+	}
+	lockStartTime := model.TSConvert2Time(lc.StartTS)
+	currentTime := model.TSConvert2Time(currentStartTS)
+	expired := currentTime.Sub(lockStartTime) > d.lease
+	if expired {
+		logutil.Logger(d.outerCtx).Warn("lock is expired",
+			zap.Time("lockStartTime", lockStartTime),
+			zap.Time("currentTime", currentTime),
+			zap.Duration("lease", d.lease))
+	}
+	return expired
+}
+
+// Unlock unlocks the distributed lock.
+func (d *DistributedLock) Unlock() error {
+	defer func() {
+		if d.cancelRenewLoop != nil {
+			d.cancelRenewLoop()
+		}
+	}()
+	return d.getLockInTxnWithRetry(context.Background(),
+		func(lc *lockContent, txn DataTxn) (storeErr, otherErr error) {
+			if lc == nil || lc.InstanceID != d.instanceID {
+				logutil.Logger(d.outerCtx).Warn("lock is unexpectedly removed", zap.String("instanceID", d.instanceID))
+				return nil, nil
+			}
+
+			// Lock is exist, remove it.
+			err := txn.Del(d.lockName)
+			return errors.Trace(err), nil
+		})
+}
+
+func (d *DistributedLock) getLockInTxnWithRetry(ctx context.Context, fn lockHandleFunc) error {
+	retryCnt := 0
+	backoffTimer := time.NewTimer(d.backoff)
+	defer backoffTimer.Stop()
+	for {
+		backoffTimer.Reset(d.backoff)
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-d.outerCtx.Done():
+			// We only check the outer context error after retry 10 times,
+			// so that the lock can be released even if the outer context is canceled.
+			if retryCnt >= 10 {
+				return d.outerCtx.Err()
+			}
+		case <-backoffTimer.C:
+			var noRetry bool
```

Comment:
these logic should be moved out of `select`, and let `case <-backoffTimer.C` be an empty case. If d.outerCtx is done we have little chance to enter `case <-backoffTimer.C`